### DISCUSSION TITLE
How do I group by derived columns? Also, are common table expressions supported?

### DISCUSSION THREAD

**Discussion by grechkay** [diesel-rs/diesel#4352] [Q&A] [answered]
I'm using sqlite

Say I have a schema:

```
id INTEGER,
val INTEGER
```

I want to get the sum of the values for odd ids and even ids.

In raw sql, I would accomplish this as such:

```
WITH T1 AS (
  SELECT 
    id % 2 AS mod,
    val
)
SELECT 
  mod,
  SUM(val) AS sum_val
FROM 
  T1
GROUP BY mod
```
However I'm having a really hard time replicating this logic with the diesel query builder. Right now I am basically using raw sql for this type of logic, but if possible, I'd like to migrate it over to the query builder.

**Comment by weiznich** [comment, marked as answer]
It's always hard to answer questions that ask "Is it possible to do xyz", because almost anything is possible if you are willing to spend time implementing in.

In this case the built-in DSL does not support CTE yet, although that's something that might change in the future. That written: You likely don't need to use a CTE expression there at all as you could write that query as well as
```SQL
SELECT 
  id % 2 AS mod,
  SUM(val) AS sum_val
FROM 
  Table
GROUP BY mod
```
or as
```SQL
SELECT 
  id % 2,
  SUM(val) AS sum_val
FROM 
  Table
GROUP BY id % 2
```

The first variant would require column aliasing support in the DSL, which again is not supported yet. The second variant is technically supported, but I'm not 100% sure if that would run into #3950. 

Finally you can always extend the DSL on your own in your own crate, checkout the ["Extending Diesel"](https://diesel.rs/guides/extending-diesel.html) for details. If you really want to express it that way as DSL that's the way you should go.

**Comment by lucaoskaique** [comment]
I ran into this problem at work using diesel, i wanted to make a CTE but i couldn't find anything. I'm new to rust, but i would love to help, is there a beginner friendly PR to start gaining codebase knowledge  @weiznich ?

**Reply by weiznich** [reply]
There are unfortunately no beginner friendly taks for implementing support for CTE's as that requires first and foremost someone doing the underlying design.

If you look for other issues that are better suited for beginners I would suggest to take a look of any of [these issues](https://github.com/diesel-rs/diesel/issues?q=is%3Aissue%20state%3Aopen%20label%3A%22good%20first%20issue%22%20label%3A%22help%20wanted%22%20label%3A%22mentoring%20available%22)
